File: src/test/java/com/adinsight/agent/tools/EchoTool.java
package com.adinsight.agent.tools;

import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class EchoTool implements AgentTool<EchoTool.Input> {
    public record Input(String text) {
    }

    @Override
    public String name() {
        return "echo";
    }

    @Override
    public String description() {
        return "Echoes the provided text.";
    }

    @Override
    public Class<Input> inputType() {
        return Input.class;
    }

    @Override
    public Map<String, Object> inputSchema() {
        return Map.of(
                "type", "object",
                "properties", Map.of(
                        "text", Map.of("type", "string", "description", "Text to echo")
                ),
                "required", java.util.List.of("text")
        );
    }

    @Override
    public CompletableFuture<ToolResult> execute(Input input, ToolContext context) {
        return CompletableFuture.completedFuture(ToolResult.success(Map.of("text", input.text())));
    }
}

File: test/tools/echo-tool.ts
import { z } from "zod";
import type { Tool, JsonSchema, ToolContext } from "../../src/tools/tool.js";
import { ok, type ToolResult } from "../../src/tools/result.js";

const echoInput = z.object({
  text: z.string()
});

type EchoInput = z.infer<typeof echoInput>;

export class EchoTool implements Tool<EchoInput, Readonly<{ text: string }>> {
  readonly name = "echo";
  readonly description = "Echoes the provided text.";
  readonly input = echoInput;
  readonly parameters: JsonSchema = {
    type: "object",
    properties: {
      text: {
        type: "string",
        description: "Text to echo"
      }
    },
    required: ["text"]
  };

  async execute(
    input: EchoInput,
    _context: ToolContext
  ): Promise<ToolResult<Readonly<{ text: string }>>> {
    return ok({ text: input.text });
  }
}
